### DISCUSSION TITLE
Intercepting `eth_sendTransaction` for ERC-4337 (Smart Contract Wallets)

### DISCUSSION THREAD

**Discussion by hexcowboy** [MetaMask/snaps#1067] [Q&A] [answered]
When implementing an ERC-4337 snap, one of the issues that came up is [intercepting `eth_sendTransaction`](https://hackmd.io/@plusminushalf/metamask-snaps-4337-wallet#Sending-transaction).

Is this idea currently being explored by MM? Are there any workarounds currently? This is the only reason right now as to why we may choose to build a standalone wallet.

**Comment by ritave** [comment, marked as answer]
Hello @hexcowboy,

We want snaps to have the ability to provide support for new blockchains to users and dApps seamlessly. This will be done through a feature called "Keyrings" specified in [SIP-2](https://github.com/MetaMask/SIPs/blob/main/SIPS/sip-2.md).

It will allow a snap to specify which methods it supports and then the dApp will be directed directly to a snap, bypassing the extension itself.

We [made a custom build with partial support for this feature](https://github.com/MetaMask/snaps-monorepo/discussions/821) for Devcon 6.

We're currently focusing on getting Snaps from Flask to the main extension and work on Keyrings is currently on-hold. While I can't promise any dates we're estimating we'll be able to work on it more starting in Q2 of 2023.
